mewarpx/mewarpx/utils_store/init_restart_util.py

```diff
@@ -0,0 +1,52 @@
+"""
+Utility functions to start a run from a checkpoint or restart.
+"""
+import os
+import shutil
+import logging
+import traceback
+
+from mewarpx.mwxrun import mwxrun
+
+logger = logging.getLogger(__name__)
+
+
+def clean_old_checkpoints(prefix):
+    if os.path.isdir("diags"):
+        for d in next(os.walk("diags"))[1]:
+            if d.startswith(prefix):
+                print(f"Removing old checkpoint file {d}")
+                shutil.rmtree(os.path.join("diags", d))
+
+def run_restart(checkpoint_prefix, force=False):
+
+    logger.info(f"Attempting to " + ("force a " if force else "") + "restart from the most recent checkpoint starting with '{checkpoint_prefix}'")
+
+    if not os.path.isdir("diags"):
+        if force:
+            raise RuntimeError("Diags directory does not exist!")
+        else:
+            logger.warning("Diags directory does not exist!")
+            return
+
+    # using next() gives only the first layer of subdirectories
+    checkpoints = sorted([f for f in next(os.walk("diags"))[1] if "old" not in f and f.startswith(checkpoint_prefix)])
+
+    if not checkpoints:
+        if force:
+            raise RuntimeError(f"There were no checkpoint directories starting with {checkpoint_prefix}!")
+        else:
+            logger.warning(f"There were no checkpoint directories starting with {checkpoint_prefix}!")
+            return
+
+    try:
+        checkpoint = checkpoints[-1:][0]
+
+        logger.info(f"Restarting from {checkpoint}")
+
+        mwxrun.simulation.amr_restart = os.path.join("diags", checkpoint)
+
+    except RuntimeError:
+        logger.error("There was a error while trying to restart from the checkpoint!")
+        traceback.print_exc()
+        raise
```

Comment:
In this case we'd want to try and restart from the next checkpoint file, right? So instead of raising the exception I would say we should remove the checkpoint that failed and rerun the function. Then only raise the exception if there are no more checkpoint files left over AND `force=True` (if `force=False` the simulation should just start from the beginning).

Reply:
Actually, this will be more difficult since the restart won't actually happen here, so this `RuntimeError` won't catch failed restarts. So saying there was an error in trying to restart is not really correct, the error would be in trying to write `mwxrun.simulation.amr_restart`. We should think some more about how we would be able to catch failed restarts.

Reply:
Will we just have to do that from somewhere in the input script?

Reply:
I think we will probably need to wrap the `mwxrun.init_run()` functionality in a try statement.

Reply:
The cleanest would probably be to edit that function (`mwxrun.init_run()`) to take an argument for whether or not the run should be restarted and have it call the `init_restart_util.run_restart()` function. Could that work?

Reply:
That sounds like a clean solution, I will test that.

Reply:
@roelof-groenewald It's also worth noting we may not need the functionality of deleting restarts / trying the next one back. Ideally restarts are written to temp files and then moved with an atomic operation to the correct filename only once complete. If AMReX does this properly (I haven't checked), we won't have the issues we did with Warp where a run killed while checkpointing would have the files present but incomplete.